<?xml version="1.0"?>
<launch>
  ###########  launch config  ###########
  <arg name="real_machine" default="True" />
  <arg name="simulation" default="False" />
  <arg name="control_mode" default="0" />
  <arg name="type" default="quad" />
  <arg name="onboards_model" default="upboard_201911" />
  <arg name="estimate_mode"  default= "1" /> <!-- 2: pure mocap mode -->
  <arg name="headless" default="True" />
  <arg name="worldtype" default="$(find aerial_robot_simulation)/gazebo_model/world/empty.world" />
  <arg name="launch_gazebo" default="True" />
  <arg name="direct_model" default="False" />
  <arg name="direct_model_name" />

  ###########  UAV Config  ###########
  # DRAGON = 32 # 0x20, dragon type with full/over actuated property
  <param name="/uav_info/uav_model" value= "32" />

  ###########  Motor Config  ###########
  <rosparam file="$(find dragon)/config/MotorInfo.yaml" command="load" />

  ###########  Servo Config  ###########
  <rosparam file="$(find dragon)/config/$(arg type)/Servo.yaml" command="load" />

  ###########  Battery Config  ###########
  <rosparam file="$(find dragon)/config/$(arg type)/Battery.yaml" command="load" />

  ###########  Base Platform  ###########
  <node pkg="aerial_robot_base" type="aerial_robot_base_node" name="aerial_robot_base_node" output="screen" >

    ###########  Basic Param  ###########
    # EGOMOTION_ESTIMATE = 0
    # EXPERIMENT_ESTIMATE = 1. for unstable mocap, use this mode
    # GROUND_TRUTH = 2
    <param name="estimator/estimate_mode" value= "2" if="$(arg simulation)" />
    <param name="estimator/estimate_mode" value= "$(arg estimate_mode)" if="$(arg real_machine)"/>

    <param name="navigator/xy_control_mode"  value="$(arg control_mode)"/>

    <param name="simulation" value="$(arg simulation)" /> <!-- for sensor plugin -->
    <param name="param_verbose" value= "false" />
    <param name="main_rate" type="double" value="40"/>

    ###########  Sensor Fusion  ###########
    <rosparam file="$(find dragon)/config/$(arg type)/egomotion_estimation/$(arg onboards_model).yaml" command="load" if="$(arg real_machine)"/>
    <rosparam file="$(find dragon)/config/$(arg type)/egomotion_estimation/simulation.yaml" command="load" if="$(eval arg('simulation') * (1 - arg('real_machine')))"/>

    <remap from="/joint_states" to="/dragon/joint_states"/>

    ###########  Gimbal Control  ###########
    <param name="control_plugin_name" value ="control_plugin/dragon_gimbal" />
    <rosparam file="$(find dragon)/config/$(arg type)/GimbalControlConfig.yaml" command="load" />
    <remap from="/gimbals_ctrl" to="/dragon/gimbals_ctrl"/>
    <remap from="/joints_ctrl" to="/dragon/joints_ctrl"/>
    <remap from="/joint_states" to="/dragon/joint_states"/>

    ###########  Teleop  ###########
    <rosparam file="$(find dragon)/config/$(arg type)/TeleopNavigationConfig.yaml" command="load" />
  </node>

  ###########  Sensors  ###########
  <include file="$(find dragon)/launch/includes/$(arg onboards_model)/sensors.launch.xml" >
    <arg name="real_machine" value="$(arg real_machine)" />
    <arg name="simulation" value="$(arg simulation)" />
  </include >

  ###########  Robot Model  ###########
  <include file="$(find dragon)/launch/robot_model.launch">
    <arg name="type" value="$(arg type)" />
    <arg name="onboards_model" value="$(arg onboards_model)"/>
    <arg name="headless" value="$(arg headless)" />
    <arg name="need_joint_state" value="false" if ="$(eval arg('simulation') + arg('real_machine') > 0)"/>
    <arg name="description_mode" value="gazebo" if="$(arg simulation)"/>
    <arg name="direct_model" value="$(arg direct_model)" />
    <arg name="direct_model_name" value="$(arg direct_model_name)" if="$(arg direct_model)"/>
  </include >

  ###########  Servo Bridge  ###########
  <node pkg="aerial_robot_model" type="servo_bridge_node" name="servo_bridge"  output="screen"  ns="dragon" />

  ###########  Transform Control  ###########
  <node pkg="dragon" type="dragon_transform_control_node" name="dragon"  output="screen" respawn="true">
    <rosparam file="$(find dragon)/config/$(arg type)/AttitudeControl.yaml" command="load" />
    <remap from="joint_states" to="dragon/joint_states"/>
    <param name="kinematic_verbose" value="false" />
    <param name="control_verbose" value="false" />
    <param name="debug_verbose" value="false" />
    <param name="real_machine" value="$(arg real_machine)" />
  </node>

  ########## Simulation in Gazebo #########
  <include file="$(find aerial_robot_simulation)/launch/simulation.launch" if = "$(eval arg('simulation') * (1 - arg('real_machine')))" >
    <arg name="model" default="dragon" />
    <arg name="gui" default="false" if="$(arg headless)" />
    <arg name="headless" default="$(arg headless)" />
    <arg name="launch_gazebo" default="$(arg launch_gazebo)" />
    <arg name="worldtype" value="$(arg worldtype)" />
  </include>
</launch>



